File: Practice/AllDomains/CoreCS/Algorithms/Strings/SherlockandtheValidString.cs
/*	author:@shivkrthakur  */
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
class Solution 
{
	static bool IsValid(string s)
    {
        int [] chars = new int[26];
        for(int i = 0; i < s.Length; i++) chars[s[i] - 'a']++;

        var dict = new Dictionary<int,int>();
        for(int j = 0; j < chars.Length; j++) 
            if(chars[j] > 0)
            if(dict.ContainsKey(chars[j])) dict[chars[j]]++;
			else dict.Add(chars[j], 1);

        if(dict.Count == 1) return true;
        if(dict.Count > 2) return false;
        var arr = dict.ToArray();

        if((arr[0].Value == 1 || arr[1].Value == 1) && (arr[0].Key == 1 || arr[1].Key == 1 || (Math.Abs(arr[0].Key - arr[1].Key) == 1))) return true;
        else return false;
    }

    static void Main(String[] args) {
        string s = Console.ReadLine();
        Console.WriteLine(IsValid(s) ? "YES" : "NO");
    }
}

File: Practice/AllDomains/CoreCS/Algorithms/Strings/SherlockandtheValidString.js
/*	author:@shivkrthakur  */
process.stdin.resume();
process.stdin.setEncoding('ascii');

var input_stdin = "";
var input_stdin_array = "";
var input_currentline = 0;

process.stdin.on('data', function (data) {
    input_stdin += data;
});

process.stdin.on('end', function () {
    input_stdin_array = input_stdin.split("\n");
    main();    
});

function readLine() {
    return input_stdin_array[input_currentline++];
}

/////////////// ignore above this line ////////////////////

function isValid(str){
    // Complete this function
      var dict = {};
      for(var j = 0; j < str.length; j++){
        if(dict[str[j]]) dict[str[j]]++;
        else dict[str[j]] = 1;
      }

      var dict2 = {};
      for(var item in dict){
          if(dict2[dict[item]])dict2[dict[item]]++;
          else dict2[dict[item]] = 1;
      }

      var flag = false;
      if(Object.keys(dict2).length <= 2){
        var keys = Object.keys(dict2);
        if(keys.length == 2){
          if((dict2[keys[0]] == 1 && (Math.abs(keys[0] - keys[1]) == 1 || keys[0] == 1)) || (dict2[keys[1]] == 1 && (Math.abs(keys[0] - keys[1]) == 1 || keys[1] == 1))){
            flag = true;
          }
          else flag = false;
        }
        else flag = true;
      }
      if(flag) return 'YES';
      else return 'NO';
}

function main() {
    var s = readLine();
    var result = isValid(s);
    process.stdout.write(""+result+"\n");

}
